### PR TITLE
Fluentd cloudwatch

### LINKED ISSUE #487: operator support for fluentd kinesis/cloudwatch plugins
### Is your feature request related to a problem? Please describe.

I have fluent operator enhanced to support using the fluentd kinesis plugin and am working on cloudwatch plugin support.  Is there interest in having this added to the main codebase, or was there a conscious decision to limit the fluentd plugins that are supported?

### Describe the solution you'd like

.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by jseutter**
### What this PR does / why we need it:
This change adds support for the cloudwatch plugin in fluentd which is able to send logs from Fluentd to AWS Cloudwatch.

Important note: This change also makes the fluentbit pod run as a privileged container so it can read from the container logs on the kubernetes node.  I'm guessing by the fact this wasn't already done that there are use cases for fluentbit that don't read from the container logs on kubernetes nodes, so should this be a configurable setting?  The code in question is in pkg/operator/daemonset.go

I tested this change in OpenShift, not sure if that is significant or not.

`make test` fails because the test output differs by a newline character at the end of the output.  Is there an easy way to fix this?

### Which issue(s) this PR fixes:
Fixes #487 

### Does this PR introduced a user-facing change?

```release-note
Added support for the fluentd cloudwatch plugin to send logs to AWS Cloudwatch
```

### Additional documentation, usage docs, etc.:

```docs

```

### CODE REVIEW

File: apis/fluentd/v1alpha1/plugins/output/firehose.go [hunk 1]
```diff
@@ -0,0 +1,31 @@
+package output
+
+// Firehose defines the parametes for out_firehose output plugin
+type Firehose struct {
```
Comment (wenchajun) [inline apis/fluentd/v1alpha1/plugins/output/firehose.go:4]: Why is it firehose here?

Comment (jseutter) [reply apis/fluentd/v1alpha1/plugins/output/firehose.go]: Oops, removing..

File: pkg/operator/daemonset.go [hunk 2]
```diff
@@ -80,6 +80,9 @@ func MakeDaemonSet(fb fluentbitv1alpha2.FluentBit, logPath string) appsv1.Daemon
 							},
 							ReadinessProbe: fb.Spec.ReadinessProbe,
 							LivenessProbe:  fb.Spec.LivenessProbe,
+							SecurityContext: &corev1.SecurityContext{
+								Privileged: func(b bool) *bool { return &b }(true),
+							},
```
Comment (benjaminhuo) [inline pkg/operator/daemonset.go:85]: @wanjunlei I think we needn't make fluentbit pod privileged pod to read logs from k8s node?

File: apis/fluentd/v1alpha1/plugins/output/types.go [hunk 3]
```diff
@@ -42,6 +42,8 @@ type Output struct {
 	Stdout *Stdout `json:"stdout,omitempty"`
 	// out_loki plugin
 	Loki *Loki `json:"loki,omitempty"`
+	// out_cloudwatch plugin
+	Cloudwatch *Cloudwatch `json:"cloudwatch,omitempty"`
```
Comment (benjaminhuo) [inline apis/fluentd/v1alpha1/plugins/output/types.go:46]: @jseutter would you help to take a look at the github action errors?

**Comment by wenchajun** [comment]
You should add it here. https://github.com/fluent/fluent-operator/tree/master/apis/fluentd/v1alpha1/plugins/output
It looks like you don't have a Cloudwatch structure defined.

File: apis/fluentd/v1alpha1/plugins/output/types.go (hunk 3 above)
Comment (jseutter) [reply apis/fluentd/v1alpha1/plugins/output/types.go:46]: The missing definition error should be fixed now.

**Comment by benjaminhuo** [comment]
@jseutter Unit test failed:
image

**Comment by jseutter** [comment]
This is the error I referenced in the PR description above.  The difference between the expected and actual outputs is a newline character.  Do you know how to control the output to not emit the newline?  Nothing I have tried has worked.

**Comment by wanjunlei** [comment]
> 

Maybe you can delete the `\n` before comparing.

File: pkg/operator/daemonset.go (hunk 2 above)
Comment (wanjunlei) [reply pkg/operator/daemonset.go:85]: Yes, the fluentbit pod should not be a privileged pod by default.

If you want to make fluentbit pod a privileged pod, please set a `SecurityContext` in [FlunetBit CRD](https://github.com/fluent/fluent-operator/blob/master/apis/fluentbit/v1alpha2/fluentbit_types.go#L73).

**Comment by jseutter** [comment]
*smacks forehead* - great idea, I'll try that.  Thanks!

File: pkg/operator/daemonset.go (hunk 2 above)
Comment (jseutter) [reply pkg/operator/daemonset.go:85]: I'll test this tomorrow to figure out what OpenShift was preventing the unprivileged pod from doing.

Comment (benjaminhuo) [reply pkg/operator/daemonset.go:85]: > If you want to make fluentbit pod a privileged pod, please set a `SecurityContext` in [FlunetBit CRD](https://github.com/fluent/fluent-operator/blob/master/apis/fluentbit/v1alpha2/fluentbit_types.go#L73).

Right, then @jseutter I think you can revert this change to the default of setting securityContext to true.
You can set it in the FluentBit CRD if you need it

**Comment by benjaminhuo** [comment]
@jseutter We're going to release fluent operator v2.0 in a day or two, are you able to resolve all the comments and catch up with the v2.0 release?

BTW, you can submit your PR to the master branch instead.  You might need to fetch latest updates and rebase your PR
I noticed that you submit your PR to the release-1.7 branch:
image

**Comment by benjaminhuo** [comment]
@jseutter you're submitting this PR to fluent-operator's `release-1.7` branch, would you submit it to the `master` branch instead?

image

File: pkg/operator/daemonset.go (hunk 2 above)
Comment (jseutter) [reply pkg/operator/daemonset.go:85]: Hm, I don't see a way to set fluentbit pods to run as a privileged user.  The only privileged section in the FluentBit CRD is under spec.initContainers.securityContext which won't help me.  Is there a reason there is no spec.securityContext.privileged attribute?

Comment (wanjunlei) [reply pkg/operator/daemonset.go:85]: You need to add a `SecurityContext` field into `Fluentd` CRD like [this](https://github.com/fluent/fluent-operator/blob/master/apis/fluentbit/v1alpha2/fluentbit_types.go#L73).

Comment (jseutter) [reply pkg/operator/daemonset.go:85]: Hm, I'm talking about FluentBit in this case, not Fluentd.  Regardless, I modified the FluentBit CRD to allow spec.securityContext.privileged but the operator code won't include that setting when creating the StatefulSet.  I'll dig into the operator code next.

Comment (wanjunlei) [reply pkg/operator/daemonset.go:85]: Sorry, my mistake.

The StatefulSet is controlled by the [Collector](https://github.com/fluent/fluent-operator/blob/master/apis/fluentbit/v1alpha2/collector_types.go#L31) CRD, not the `FluentBit` CRD.

Comment (jseutter) [reply pkg/operator/daemonset.go]: I think I figured it out.  FluentBit links to a PodSecurityContext which does not have a privileged field.  There is a SecurityContext which does have a privileged field, and is used for container security contexts.  I figured out a workaround which is to use the seLinuxOptions and set the security context:

`securityContext:
  seLInuxOptions:
    level: s0
    role: system_r
    type: spc_t
    user: system_u`

Doing this it doesn't matter if the container user does not have access to the files.  I have removed the offending code from the PR.

**Pull request closed without merging**